### PR TITLE
FEAT - add fit_intercept support for LBFGS

### LINKED ISSUE #325: FEAT - Add fit_intercept support to LBFGS solver
## Description of the feature

The LBFGS solver currently doesn't support `fit_intercept=True`. A use case of this was described in Issue #320 

## Considered alternatives

Manual data centering works but complicates the optimization and user experience. In many use cases sklearn is currently is stil the faster alternative. 

**Additional context**

```
np.random.seed(42)
X = np.random.randn(1000, 50)
y = np.random.poisson(np.exp(X.sum(axis=1) * 0.01 - 1))

# This should work but currently fails as its not supported
estimator = GeneralizedLinearEstimator(
    datafit=Poisson(),
    penalty=L2(alpha=0.01),
    solver=LBFGS(fit_intercept=True)  # <- Currently not supported
)
estimator.fit(X, y)

# Compare with sklearn (works fine)
sklearn_model = PoissonRegressor(alpha=0.01, fit_intercept=True, solver='lbfgs')
sklearn_model.fit(X, y)
```

Note: Pattern exists in `anderson_cd.py` (lines 163-166) and other solvers to see how to implement fit_intercept method.

### PR DESCRIPTION
**Pull request by floriankozikowski**
## Context of the PR

Closes #325 

## Contributions of the PR

- add fit_intercept support for LBFGS 
- add default warm_start=False, so LBFGS works with debug script in #320

### Checks before merging PR

- [ ] added documentation for any new feature
- [ ] added unit tests
- [ ] edited the [what's new](../doc/changes/whats_new.rst) (if applicable)

### CODE REVIEW

File: skglm/solvers/lbfgs.py [hunk 1]
```diff
@@ -46,25 +51,46 @@ def _solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
...
-            datafit_value = datafit.value(y, w, Xw)
-            penalty_value = penalty.value(w)
+            if self.fit_intercept:
+                Xw = X @ w[:-1] + w[-1]
+                datafit_value = datafit.value(y, w[:-1], Xw)
+                penalty_value = penalty.value(w[:-1])
+            else:
+                Xw = X @ w
+                datafit_value = datafit.value(y, w, Xw)
+                penalty_value = penalty.value(w)
```
Comment (Badr-MOUFAD) [inline skglm/solvers/lbfgs.py:61]: What do you think about this refactoring @floriankozikowski ?
```suggestion
            w_features = w[:n_features]
            Xw = X @ w_features
            if self.fit_intercept:
                Xw += w[-1]
                
            datafit_value = datafit.value(y, w_features , Xw)
            penalty_value = penalty.value(w_features)
```

File: skglm/solvers/lbfgs.py [hunk 2]
```diff
@@ -46,25 +51,46 @@ def _solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
...
-            Xw = X @ w
-            datafit_grad = datafit.gradient(X, y, Xw)
-            penalty_grad = penalty.gradient(w)
-
-            return datafit_grad + penalty_grad
+            if self.fit_intercept:
+                Xw = X @ w[:-1] + w[-1]
+                datafit_grad = datafit.gradient(X, y, Xw)
+                penalty_grad = penalty.gradient(w[:-1])
+                intercept_grad = datafit.intercept_update_step(y, Xw)
```
Comment (mathurinm) [inline skglm/solvers/lbfgs.py:70]: there may be an issue here, because intercept_update_step does not compute the gradient, but a scaled version of it (it's mulliplied by the stepsize).
The safest way to do it would be to call `raw_grad().sum()`, which is equivalent to `np.ones(n_features) @ raw_grad()`, e.g. the gradient with respect to a feature full of ones.

Sounds good @Badr-MOUFAD ?

**Comment by floriankozikowski** [comment]
@Badr-MOUFAD I tried the refactor (also considering your comment @mathurinm )
Let me know what you think. Didn't find an option to make this refactor shorter

**Comment by floriankozikowski** [comment]
Btw, looking at the initial issue #320 that initiated this PR, the intercept fitting improved the sklearn speed difference, but it is still slower. 

--- Fitting Time Comparison ---
skglm (LBFGS):   8.0444 seconds
sklearn (L-BFGS):     2.4578 seconds
 >> sklearn was 3.27x faster.

I guess in this PR, we only focus on the intercept, but maybe we should open a new issue investigating other causes for this. 

If you approve the refactor, I can delete the debug script (issue320) and we can merge.

File: skglm/solvers/lbfgs.py (hunk 2 above)
Comment (Badr-MOUFAD) [reply skglm/solvers/lbfgs.py]: Good catch @mathurinm,

I have a little concern because it seems that in some part of the code (`Logostic` datafit) `intercept_update_step` account for the stepsize, whereas in other parts (`Quadratic`, `Huber`, `Poisson`, ...) `intercept_update_step` evaluate the gradient

**Comment by Badr-MOUFAD** [comment]
Thanks @floriankozikowski for the timing comparison.
Weird that sklearn is x3 faster, but I agree to tackle this issue in a separate PR

Can you pls add the intercept to the unittest and update the whats'new page

**Comment by floriankozikowski** [comment]
@Badr-MOUFAD thanks for the feedback! It should be complete now and I'd say we can merge. I won't have access to my laptop the next two weeks, so if anything comes up, I will look at it mid-August again.
